Test that packages don't start elasticsearch

nik9000 opened the pull request:
We don't want either the deb or rpm package to start elasticsearch as soon
as they install nor do we want the package to register elasticsearch to
start on restart. That action is reserved for the administrator. This adds
tests for that.

Closes #13122

nik9000 committed:
[test] Test that packages don't start service

We don't want either the deb or rpm package to start elasticsearch as soon
as they install nor do we want the package to register elasticsearch to
start on restart. That action is reserved for the administrator. This adds
tests for that.

Closes #13122

nik9000 commented:
Ping @tlrx more bats to review! This one verifies that we're consistent across the rpm and deb on starting Elasticsearch. We've had some talks on github about whether the deb should start elasticsearch when it installs it and we came down on that being a bad idea. Its agains debian policy, but we're in good company in not liking that policy.

dakrone commented on qa/vagrant/src/test/resources/packaging/scripts/30_deb_package.bats:
```diff
@@ -63,16 +63,17 @@ setup() {
+    # Wait a second to give Elasticsearch a change to start if it is going to.
+    # This isn't perfect by any means but its something.
+    sleep 1
+    ! ps aux | grep elasticsearch | grep java
```
What about using `! jps | grep -F -i elasticsearch`

nik9000 replied:
I'll give it a shot!

nik9000 replied:
Ok - gave it a try. You need: ```! sudo -u elasticsearch jps | grep -i elasticsearch``` which is pretty cumbersome and seems more error prone so I'm going to stick with `ps aux`

dakrone commented on qa/vagrant/src/test/resources/packaging/scripts/70_sysv_initd.bats:
```diff
@@ -37,11 +37,28 @@ setup() {
+@test "[INIT.D] daemon isn't enabled on restart" {
+    # Rather than restart the VM which would be slow we check for the symlinks
+    # that init.d uses to restart the application on startup.
+    ! find /etc/rc[0123456].d | grep elasticsearch
```
can be `! find /etc/rc[0123456].d -iname "elasticsearch*"`

nik9000 replied:
Yeah - that is probably a bit better. Will try it.

nik9000 replied:
And that'd look like: `[ $(find /etc/rc[0123456].d -iname "elasticsearch*" | wc -l) -eq 0 ]` which isn't really nicer either!

dakrone commented:
Left some minor comments, otherwise LGTM

drewr commented:
I absolutely :+1: this, glad that it's finally fixed

nik9000 committed:
[test] Document implementation choices

There are two other obvious ways to implement the "packages don't start
elasticsearch" checks but when you work through them they aren't as nice
as the implementation of the checks that we use now. This just adds
documentation to that effect.

nik9000 commented:
Merged to master, 2.x, and 2.0.
